### PR TITLE
[TypeInfo] Handle custom collection objects properly

### PR DESCRIPTION
**Pull request by mtarld**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | no 
| Deprecations? | no 
| Issues        | 
| License       | MIT

Understand custom collection objects.

### CODE REVIEW

**Review by chalasr — APPROVED** [review]

**Review by welcoMattic — APPROVED** [review]

File: src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php [hunk 1]
```diff
@@ -164,7 +162,7 @@ private function getTypeFromNode(TypeNode $node, ?TypeContext $typeContext): Typ
                 default => $this->resolveCustomIdentifier($node->name, $typeContext),
             };
 
-            if ($type instanceof ObjectType && \in_array($type->getClassName(), self::COLLECTION_CLASS_NAMES, true)) {
+            if ($type instanceof ObjectType && (is_a($type->getClassName(), \Traversable::class, true) || is_a($type->getClassName(), \ArrayAccess::class, true))) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php:165]: is_subclass_of?

Comment (mtarld) [reply src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php:165]: When using `is_subclass_of`, only subclasses will return true. And I think that it'd be great to be able to understand `Traversable` as a collection as well. WDYT?

**Review by Korbeil — APPROVED** [review]
:+1:

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php:165]: Not sure what you mean. is_subclass_of works with interfaces too: https://3v4l.org/bP6JN

Comment (xabbuh) [reply src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php:165]: I guess what @mtarld means is that `is_subclass_of(\Traversable::class, \Traversable::class);` returns `false`: https://3v4l.org/1CbGV

Comment (mtarld) [reply src/Symfony/Component/TypeInfo/TypeResolver/StringTypeResolver.php:165]: Exactly, thanks @xabbuh.

**Comment by chalasr** [comment]
Thank you @mtarld.

**Pull request merged**